### PR TITLE
Add table oci_devops_project Closes #487

### LINKED ISSUE #487: Add table oci_devops_project
**References**
Add table for OCI Devops Projects 
See Terraform definition of resource https://registry.terraform.io/providers/oracle/oci/latest/docs/data-sources/devops_project

### PR DESCRIPTION
**Pull request by ParthaI**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
Test names: [ 'tests/oci_devops_project' ]
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging

SETUP: tests/oci_devops_project []

PRETEST: tests/oci_devops_project

TEST: tests/oci_devops_project
Running terraform

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # oci_devops_project.named_test_resource will be created
  + resource "oci_devops_project" "named_test_resource" {
      + compartment_id    = "ocid1.tenancy.oc1..ghtyujhghnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4hrtgfhytgrt"
      + defined_tags      = (known after apply)
      + description       = (known after apply)
      + freeform_tags     = (known after apply)
      + id                = (known after apply)
      + lifecycle_details = (known after apply)
      + name              = "steampipetest4436"
      + namespace         = (known after apply)
      + state             = (known after apply)
      + system_tags       = (known after apply)
      + time_created      = (known after apply)
      + time_updated      = (known after apply)

      + notification_config {
          + topic_id = (known after apply)
        }
    }

  # oci_ons_notification_topic.named_test_ons_topic_resource will be created
  + resource "oci_ons_notification_topic" "named_test_ons_topic_resource" {
      + api_endpoint   = (known after apply)
      + compartment_id = "ocid1.tenancy.oc1..ghtyujhghnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4hrtgfhytgrt"
      + defined_tags   = (known after apply)
      + description    = "Terraform testing resource"
      + etag           = (known after apply)
      + freeform_tags  = {
          + "Name" = "steampipetest4436"
        }
      + id             = (known after apply)
      + name           = "steampipetest4436"
      + short_topic_id = (known after apply)
      + state          = (known after apply)
      + time_created   = (known after apply)
      + topic_id       = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + notification_topic_id = (known after apply)
  + reporting_region      = "us-ashburn-1"
  + resource_id           = (known after apply)
  + resource_name         = "steampipetest4436"
  + tenancy_ocid          = "ocid1.tenancy.oc1..ghtyujhghnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4hrtgfhytgrt"
oci_ons_notification_topic.named_test_ons_topic_resource: Creating...
oci_ons_notification_topic.named_test_ons_topic_resource: Creation complete after 2s [id=ocid1.onstopic.oc1.iad.aaaaaaaa4qnu4njeetkvh2k5pfyy66f5nhqvep3qy6vi32k34t4e3cbeqtkq]
oci_devops_project.named_test_resource: Creating...
oci_devops_project.named_test_resource: Creation complete after 3s [id=ocid1.devopsproject.oc1.iad.amaaaaaa6igdexaatbewxbpch6mpbzqjpamtgd7ciyu3h44y4a3gabiibtba]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.

Outputs:

notification_topic_id = "ocid1.onstopic.oc1.iad.aaaaaaaa4qnu4njeetkvh2k5pfyy66f5nhqvep3qy6vi32k34t4e3cbeqtkq"
reporting_region = "us-ashburn-1"
resource_id = "ocid1.devopsproject.oc1.iad.amaaaaaa6igdexaatbewxbpch6mpbzqjpamtgd7ciyu3h44y4a3gabiibtba"
resource_name = "steampipetest4436"
tenancy_ocid = "ocid1.tenancy.oc1..ghtyujhghnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4hrtgfhytgrt"

Running SQL query: test-get-query.sql

[
  {
    "freeform_tags": {},
    "id": "ocid1.devopsproject.oc1.iad.amaaaaaa6igdexaatbewxbpch6mpbzqjpamtgd7ciyu3h44y4a3gabiibtba",
    "lifecycle_state": "ACTIVE",
    "name": "steampipetest4436"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql

[
  {
    "id": "ocid1.devopsproject.oc1.iad.amaaaaaa6igdexaatbewxbpch6mpbzqjpamtgd7ciyu3h44y4a3gabiibtba",
    "name": "steampipetest4436"
  }
]
✔ PASSED

Running SQL query: test-not-found-query.sql

null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "tenant_id": "ocid1.tenancy.oc1..ghtyujhghnm7gleh5soecxzjetci3yjjnjqmfkr4po3hoz4p4hrtgfhytgrt",
    "title": "steampipetest4436"
  }
]
✔ PASSED

TEARDOWN: tests/oci_devops_project

SUMMARY:

1/1 passed.
```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
```



> select * from oci_devops_project
+--------+-------------+--------------------------------------------------------------------------------------------------+---------------------------+---------------------------+-------------------+-----------------+--------------+------------------------------------->
| name   | description | id                                                                                               | time_created              | time_updated              | lifecycle_details | lifecycle_state | namespace    | notification_topic_id               >
+--------+-------------+--------------------------------------------------------------------------------------------------+---------------------------+---------------------------+-------------------+-----------------+--------------+------------------------------------->
| test32 |             | ocid1.devopsproject.oc1.ap-mumbai-1.amaaaaaa6igdexaaomriwmluck7tsrspimuet2x5dqz6m225p2hnzarxff4q | 2023-06-23T17:21:25+05:30 | 2023-06-23T17:22:00+05:30 | <null>            | ACTIVE          | bmqeqvslavsz | ocid1.onstopic.oc1.ap-mumbai-1.aaaaa>
+--------+-------------+--------------------------------------------------------------------------------------------------+---------------------------+---------------------------+-------------------+-----------------+--------------+------------------------------------->

Time: 4.2s. Rows fetched: 1. Hydrate calls: 1.
> .cache off
> .cache clear
> select * from oci_devops_project where id = 'ocid1.devopsproject.oc1.ap-mumbai-1.amaaaaaa6igdexaaomriwmluck7tsrspimuet2x5dqz6m225p2hnzarxff4q'
+--------+-------------+--------------------------------------------------------------------------------------------------+---------------------------+---------------------------+-------------------+-----------------+--------------+------------------------------------->
| name   | description | id                                                                                               | time_created              | time_updated              | lifecycle_details | lifecycle_state | namespace    | notification_topic_id               >
+--------+-------------+--------------------------------------------------------------------------------------------------+---------------------------+---------------------------+-------------------+-----------------+--------------+------------------------------------->
| test32 |             | ocid1.devopsproject.oc1.ap-mumbai-1.amaaaaaa6igdexaaomriwmluck7tsrspimuet2x5dqz6m225p2hnzarxff4q | 2023-06-23T17:21:25+05:30 | 2023-06-23T17:22:00+05:30 |                   | ACTIVE          | bmqeqvslavsz | ocid1.onstopic.oc1.ap-mumbai-1.aaaaa>
+--------+-------------+--------------------------------------------------------------------------------------------------+---------------------------+---------------------------+-------------------+-----------------+--------------+------------------------------------->
```
</details>

### CODE REVIEW

File: docs/tables/oci_devops_project.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+A DevOps project is created to successfully build and deploy applications by using the DevOps service. A project logically groups the DevOps resources needed to implement a CI/CD workflow.
+
+## Examples
+
+### Basic info
+
+```sql
+select
+  display_name,
```
Comment (madhushreeray30) [inline docs/tables/oci_devops_project.md:11]:
```suggestion
  name,
```

File: docs/tables/oci_devops_project.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  id,
+  description,
+  namespace,
+  state,
```
Comment (madhushreeray30) [inline docs/tables/oci_devops_project.md:15]:
```suggestion
  lifecycle_state,
```

File: docs/tables/oci_devops_project.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  project.id,
+  project.notification_topic_id,
+  topic.name notification_topic_name,
+  project.namespace,
+  project.description
+from
+  oci_devops_project project
+  left join
+  oci_ons_notification_topic topic
+  on project.notification_topic_id = topic.topic_id
```
Comment (madhushreeray30) [inline docs/tables/oci_devops_project.md:54]:
```suggestion
  left join oci_ons_notification_topic topic on project.notification_topic_id = topic.topic_id;
```

File: oci/table_oci_devops_project.go [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+			{
+				Name:        "defined_tags",
+				Description: ColumnDescriptionDefinedTags,
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "freeform_tags",
+				Description: ColumnDescriptionFreefromTags,
+				Type:        proto.ColumnType_JSON,
+			},
```
Comment (madhushreeray30) [inline oci/table_oci_devops_project.go:103]: Add column for `system_tags`

File: oci/table_oci_devops_project.go [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+		topicId = *d.HydrateItem.(devops.ProjectSummary).NotificationConfig.TopicId
+	case devops.Project:
+		topicId = *d.HydrateItem.(devops.Project).NotificationConfig.TopicId
+	}
+	return topicId, nil
+}
+
+// Priority order for tags
+// 1. Defined Tags
+// 2. Free-form tags
```
Comment (madhushreeray30) [inline oci/table_oci_devops_project.go:277]:
```suggestion
// Priority order for tags
// 1. System Tags
// 2. Defined Tags
// 3. Free-form tags
```

File: oci/table_oci_devops_project.go [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+func devopsProjectTags(ctx context.Context, d *transform.TransformData) (interface{}, error) {
+	var freeFormTags map[string]string
+	var definedTags map[string]map[string]interface{}
```
Comment (madhushreeray30) [inline oci/table_oci_devops_project.go:280]:
```suggestion
	var definedTags map[string]map[string]interface{}
	var systemTags map[string]map[string]interface{}
```

File: oci/table_oci_devops_project.go [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+			tags = map[string]interface{}{}
+		}
+		for _, v := range definedTags {
+			for key, value := range v {
+				tags[key] = value
+			}
+
+		}
+	}
+
```
Comment (madhushreeray30) [inline oci/table_oci_devops_project.go:311]:
```suggestion
if systemTags != nil {
		if tags == nil {
			tags = map[string]interface{}{}
		}
		for _, v := range systemTags {
			for key, value := range v {
				tags[key] = value
			}

		}
	}
```

File: oci/table_oci_devops_project.go [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+// Build additional filters
+func buildProjectFilters(equalQuals plugin.KeyColumnEqualsQualMap, logger hclog.Logger) (devops.ListProjectsRequest, bool) {
+	request := devops.ListProjectsRequest{}
+	isValid := true
+
+	if equalQuals["name"] != nil && strings.Trim(equalQuals["name"].GetStringValue(), " ") != "" {
+		request.Name = types.String(equalQuals["name"].GetStringValue())
+	}
+	if equalQuals["lifecycle_state"] != nil && strings.Trim(equalQuals["lifecycle_state"].GetStringValue(), " ") != "" {
```
Comment (madhushreeray30) [inline oci/table_oci_devops_project.go:258]: Should we also add lifecycle_state in the optional quals in listconfig?

File: docs/tables/oci_devops_project.md
Comment (madhushreeray30) [inline docs/tables/oci_devops_project.md:1]: Please add more example queries, few suggestions would be 
1. List projects which are not active
2. List projects created in the last 30 days

**Review by madhushreeray30 — COMMENTED** [review]
@ParthaI can you please have a look at the comments, thanks!

**Comment by misraved** [comment]
@lucasjellema we have addressed the review comments of https://github.com/turbot/steampipe-plugin-oci/pull/489 in this PR.

Apologies for the delayed response and thanks a lot for providing the initial code structure and the insightful table docs 👍.

File: oci/table_oci_devops_project.go [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+// 2. Defined Tags
+// 3. Free-form tags
+func devopsProjectTags(ctx context.Context, d *transform.TransformData) (interface{}, error) {
+	var freeFormTags map[string]string
+	var definedTags map[string]map[string]interface{}
+	var systemTags map[string]map[string]interface{}
+	switch d.HydrateItem.(type) {
+	case devops.ProjectSummary:
+		freeFormTags = d.HydrateItem.(devops.ProjectSummary).FreeformTags
+		definedTags = d.HydrateItem.(devops.ProjectSummary).DefinedTags
```
Comment (madhushreeray30) [inline oci/table_oci_devops_project.go:288]:
```suggestion
		definedTags = d.HydrateItem.(devops.ProjectSummary).DefinedTags
		systemTags = d.HydrateItem.(devops.ProjectSummary).SystemTags
```

File: oci/table_oci_devops_project.go [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+	case devops.Project:
+		freeFormTags = d.HydrateItem.(devops.Project).FreeformTags
+		definedTags = d.HydrateItem.(devops.Project).DefinedTags
```
Comment (madhushreeray30) [inline oci/table_oci_devops_project.go:291]:
```suggestion
		definedTags = d.HydrateItem.(devops.Project).DefinedTags
		systemTags = d.HydrateItem.(devops.Project).SystemTags
```

**Review by misraved — APPROVED** [review]

**Pull request merged**